runtime/io/common/src/aws/smithy/kotlin/runtime/io/SdkBufferedSink.kt

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package aws.smithy.kotlin.runtime.io
+
+/**
+ * A sink that keeps a buffer internally so that callers can do small writes without
+ * a performance penalty.
+ */
+public expect sealed interface SdkBufferedSink : SdkSink {
+    /**
+     * The underlying buffer for this sink
+     */
+    public val buffer: SdkBuffer
+
+    /**
+     * Write [limit] bytes from [source] starting at [offset]
+     */
+    public fun write(source: ByteArray, offset: Int = 0, limit: Int = source.size - offset): Unit
+
+    /**
+     * Write all bytes from [source] to this sink.
+     * @return the number of bytes read which will be 0 if [source] is exhausted
+     */
+    public fun writeAll(source: SdkSource): Long
+
+    /**
+     * Removes [byteCount] bytes from [source] and writes them to this sink.
+     */
+    public fun write(source: SdkSource, byteCount: Long): Unit
+
+    /**
+     * Write UTF8-bytes of [string] to this sink starting at [start] index up to [endExclusive] index.
+     */
+    public fun writeUtf8(string: String, start: Int = 0, endExclusive: Int = string.length): Unit
```

Comment:
**Question**: Kotlin common seemingly only supports UTF-8 and none of the resulting methods (e.g., `String.encodeToByteArray`) mention UTF-8 in the name, merely the documentation. Is there a reason to do so here? Or could this be cleaned up by just calling it `write` or possibly `writeString`? (Same question on source read method...)

Reply:
I actually like the charset name in the function, it's very clear what is going to happen. That being said I don't feel strongly one way or the other.